### PR TITLE
new DenseKMeans using ML Dataset API

### PR DESCRIPTION
**Pull request by xwu99**
new DenseKMeans implementation with ml Dataset API as mllib RDD API has been obsoleted. 
Only works on Spark 2.4.0+ and Spark 3.0.0+

### CODE REVIEW

**Comment by xwu99** [comment]
@carsonwang CI failed due to not using Spark 3.0 in the CI script. Need to add spark 3.0 support first.

mvn clean package -q -Dmaven.javadoc.skip=true -Dspark=2.2 -Dscala=2.11
[ERROR] /home/travis/build/Intel-bigdata/HiBench/sparkbench/ml/src/main/scala/com/intel/sparkbench/ml/DenseKMeans.scala:115: error: value trainingCost is not a member of org.apache.spark.ml.clustering.KMeansSummary
[ERROR]     val cost = model.summary.trainingCost

**Comment by xwu99** [comment]
@carsonwang @gczsjdy rebased to master

**Comment by gczsjdy** [comment]
What happens if we run this new ml Dataset API KMeans with Spark 2.3/2.2/2.1?

File: sparkbench/ml/src/main/scala/com/intel/sparkbench/ml/MLDenseKMeans.scala [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+import org.apache.hadoop.io.LongWritable
+import org.apache.mahout.math.VectorWritable
+import org.apache.spark.ml.clustering.KMeans
+import org.apache.spark.ml.linalg.Vectors
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.storage.StorageLevel
+import scopt.OptionParser
+
+object DenseKMeans {
```
Comment (gczsjdy) [inline sparkbench/ml/src/main/scala/com/intel/sparkbench/ml/MLDenseKMeans.scala:28]: Add comments here to differentiate with the mllib kmeans?

File: sparkbench/ml/pom.xml [hunk 2]
```diff
@@ -54,4 +54,194 @@
...
+        <spark.bin.version>3.0</spark.bin.version>
+      </properties>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>net.alchim31.maven</groupId>
+            <artifactId>scala-maven-plugin</artifactId>
+            <configuration>
+              <excludes>
+                <exclude>com/intel/sparkbench/ml/DenseKMeans.scala</exclude>
```
Comment (gczsjdy) [inline sparkbench/ml/pom.xml:233]: Is this really necessary? I think at least this need to be clearly documented, mentioning when RDD KMeans/Dataset KMeans is leveraged.

Another option is we just remove the mllib KMeans from this release, only keeping the ml version. This approach makes more sense to me. cc @carsonwang

Comment (carsonwang) [reply sparkbench/ml/pom.xml:233]: yes, it is fine to remove Spark 2.2 2.3 support and only keep the ML version for Spark 2.4 and 3.x.

Comment (xwu99) [reply sparkbench/ml/pom.xml]: OK. it's simpler. I will revert the last commit.

**Comment by xwu99** [comment]
> What happens if we run this new ml Dataset API KMeans with Spark 2.3/2.2/2.1?

There are some parameters not supported by old versions.

**Comment by xwu99** [comment]
@gczsjdy @minmingzhu CI failed on building with old versions. Pls help to remove old Spark versions (2.0-2.3) in CI.

**Comment by gczsjdy** [comment]
I have merged https://github.com/Intel-bigdata/HiBench/pull/637, could you retrigger Travis?

**Comment by carsonwang** [comment]
need to merge master

**Comment by xwu99** [comment]
> need to merge master

done.

**Comment by shagun971** [comment]
> @carsonwang CI failed due to not using Spark 3.0 in the CI script. Need to add spark 3.0 support first.
> 
> mvn clean package -q -Dmaven.javadoc.skip=true -Dspark=2.2 -Dscala=2.11
> [ERROR] /home/travis/build/Intel-bigdata/HiBench/sparkbench/ml/src/main/scala/com/intel/sparkbench/ml/DenseKMeans.scala:115: error: value trainingCost is not a member of org.apache.spark.ml.clustering.KMeansSummary
> [ERROR] val cost = model.summary.trainingCost

How to solve this error? I am getting error when building spark 2.2 version

**Comment by gczsjdy** [comment]
@shagun971 HiBench from 8.0-SNAPSHOT doesn't support spark 2.2-, I will later update the docs to clarify that. So you can either
- Use HiBench 7.1 or former version or
- Remove the compile erroed classes in master branch manually

**Pull request merged**